packages-exp/auth-exp/src/core/mfa/mfa_error.ts

```diff
@@ -0,0 +1,68 @@
+/**
+ * @license
+ * Copyright 2020 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import * as externs from '@firebase/auth-types-exp';
+import { FirebaseError } from '@firebase/util';
+import { Auth } from '../../model/auth';
+import { IdTokenResponse } from '../../model/id_token';
+import { AuthErrorCode } from '../errors';
+import { User } from '../../model/user';
+import { AuthCredential } from '../credentials';
+import { IdTokenMfaResponse } from '../../api/authentication/mfa';
+
+export class MultiFactorError extends FirebaseError
+  implements externs.MultiFactorError {
+  readonly name = 'FirebaseError';
+  readonly code: string;
+  readonly appName: string;
+  readonly serverResponse: IdTokenMfaResponse;
+
+  readonly email?: string;
+  readonly phoneNumebr?: string;
+  readonly tenantid?: string;
+
+  constructor(
+    auth: Auth,
+    error: FirebaseError,
+    readonly operationType: externs.OperationType,
+    readonly credential?: externs.AuthCredential,
+    readonly user?: User
+  ) {
+    super(error.code, error.message);
+    // https://github.com/Microsoft/TypeScript-wiki/blob/master/Breaking-Changes.md#extending-built-ins-like-error-array-and-map-may-no-longer-work
+    Object.setPrototypeOf(this, MultiFactorError.prototype);
```

Comment:
Yeah this is funky

What about creating a type "MultiFactorError" which extends `FirebaseError`

```ts
export interface MultiFactorError extends FirebaseError {
  // Your types
}
```

Then, create a helper for example:
```ts
makeMfaError(props...) {
  const err = AuthErrorFactory.create(...) as MultiFactorError;
  Object.assign(err, props);
}
```

Reply:
you need this any time you inherit from Error, FirebaseError has to do this too

Reply:
That's what I'm saying though, just use the normal `FirebaseError`, but add typings to it so that it works in typescript

Reply:
discussed IRL, having it be MultiFactorError aligns with public documentation and allows instanceof checks